### PR TITLE
Add simulation test for Oracle module

### PR DESCRIPTION
**Pull request by RogerKSI**
## Implementation details
- Add simulation test for oracle module
- Adjust testing structure to separate testdata and testapp (prevent cycle import)
- Add Github action for running simualtion test

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [x] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

**Comment by colmazia** [comment]
Interesting, would be great to see some demos

**Review by colmazia — COMMENTED** [review]
Submit a review

File: Makefile [hunk 1]
```diff
@@ -156,4 +159,25 @@ proto-update-deps:
 ## Issue link: https://github.com/confio/ics23/issues/32
 	@sed -i '4ioption go_package = "github.com/confio/ics23/go";' $(CONFIO_TYPES)/proofs.proto
 
-.PHONY: proto-all proto-gen proto-gen-any proto-swagger-gen proto-format proto-lint proto-check-breaking proto-update-deps
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply Makefile]: added in the bottom.

File: x/oracle/simulation/operations.go [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+func SimulateMsgRequestData(
+	ak types.AccountKeeper,
+	bk simulation.BankKeeper,
+	sk types.StakingKeeper,
+	keeper keeper.Keeper,
+) simtypes.Operation {
+	return func(r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simtypes.Account, chainID string) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+		simAccount, _ := simtypes.RandomAcc(r, accs)
+
+		oCount := keeper.GetOracleScriptCount(ctx)
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply x/oracle/simulation/operations.go:147]: added

File: x/oracle/simulation/operations_test.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
+package simulation_test
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply x/oracle/simulation/operations_test.go:1]: added

**Review by colmazia — APPROVED** [review]

File: x/oracle/simulation/operations_test.go [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+	suite.ctx = app.BaseApp.NewContext(false, tmproto.Header{ChainID: "BANDCHAIN"})
+	s := rand.NewSource(1)
+	suite.r = rand.New(s)
+	suite.accs = suite.getTestingAccounts(suite.r, 10)
+
+	// begin a new block
+	suite.app.BeginBlock(
+		abci.RequestBeginBlock{
+			Header: tmproto.Header{
+				ChainID: "BANDCHAIN",
```
Comment (satawatnack) [inline x/oracle/simulation/operations_test.go:42]: "BANDCHAIN" const string?

Comment (RogerKSI) [reply x/oracle/simulation/operations_test.go]: updated

**Review by satawatnack — APPROVED** [review]

File: x/oracle/simulation/genesis.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
+package simulation
+
+// DONTCOVER
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply x/oracle/simulation/genesis.go]: removed

File: x/oracle/abci.go [hunk 6]
```diff
@@ -11,8 +11,11 @@ import (
 // handleBeginBlock re-calculates and saves the rolling seed value based on block hashes.
 func handleBeginBlock(ctx sdk.Context, req abci.RequestBeginBlock, k keeper.Keeper) {
 	// Update rolling seed used for pseudorandom oracle provider selection.
-	rollingSeed := k.GetRollingSeed(ctx)
-	k.SetRollingSeed(ctx, append(rollingSeed[1:], req.GetHash()[0]))
+	hash := req.GetHash()
+	if len(hash) > 0 {
+		rollingSeed := k.GetRollingSeed(ctx)
+		k.SetRollingSeed(ctx, append(rollingSeed[1:], hash[0]))
+	}
```
[the remark this answers is not in the document]
Comment (RogerKSI) [reply x/oracle/abci.go:19]: added comment. In the simulation test, it's possible to have empty hash on the first block.

File: x/oracle/abci.go [hunk 7]
```diff
@@ -11,8 +11,12 @@ import (
 // handleBeginBlock re-calculates and saves the rolling seed value based on block hashes.
 func handleBeginBlock(ctx sdk.Context, req abci.RequestBeginBlock, k keeper.Keeper) {
 	// Update rolling seed used for pseudorandom oracle provider selection.
-	rollingSeed := k.GetRollingSeed(ctx)
-	k.SetRollingSeed(ctx, append(rollingSeed[1:], req.GetHash()[0]))
+	hash := req.GetHash()
+	// On the firsrt block in the test. it's possible to have empty hash.
```
Comment (taobun) [inline x/oracle/abci.go:15]: first

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**